### DISCUSSION TITLE
RACK 7.0 - SPARQLGraph error:  "only recognizes nodegroup json up to version..."

### DISCUSSION THREAD

**Discussion by Robert-Adelard** [ge-high-assurance/RACK#453] [Q&A] [answered]
If you try and load a nodegroup from SPARQLGraph, you get the following error message:

```
Error: Error: belmont.js:SemanticNodeGroup.addJson() only recognizes nodegroup json up to version 12 but file is version 13
url: http://rack.office.adelard.com:8080/sparqlGraph/js/belmont.js
line: 1771
undefined
column: 181
error: Error: belmont.js:SemanticNodeGroup.addJson() only recognizes nodegroup json up to version 12 but file is version 13
```

**Comment by cuddihyge** [comment, marked as answer]
This means your javascript is old.
It most likely requires a  clearing of your cache (or hard-reload the SPARQLgraph page).
This is different for every browser.  Take a look at [Wiki: Bypass your Cache](https://en.wikipedia.org/wiki/Wikipedia:Bypass_your_cache)

If you are doing some kind of piecemeal download and running (i.e. not just running everything out of one Docker image) then it is possible you need to actually pull all the latest code, as things are out of sync.   This will also require a browser refresh.

We have searched but have not found a way to stop browsers to stop caching old javascript.  The solution is different for every browser.

**Comment by cuddihyge** [comment]
I would like to convert this to a discussion with a slightly cleaner title.   Discussion topics are useful as a question-answer database.  This is a recurring question that currently appears to be beyond our ability to fix any further.   @Robert-Adelard please accept the answer by clicking the checkmark, or let us know if this isn't a full answer.

**Reply by Robert-Adelard** [reply]
Thanks for your quick reply - doing a hard-reload (shift refresh) did indeed fix the problem, although it took a couple of attempts.

I have a couple of suggestions:

1. Improve the error message - for example, suggest that the user might try refreshing the page
2. Don't cache the JavaScript - for example, use a much shorter cache-expiry time

**Reply by Robert-Adelard** [reply]
This must be a common problem for anyone developing web applications, so I tried Google, which found the following article:

https://stackoverflow.com/questions/1922910/force-browser-to-clear-cache

The first idea is possibly the best - add an explicit version number to the JS files.

**Reply by cuddihyge** [reply]
@Robert-Adelard Thanks Robert.  I'll improve the error message.  We've looked at those stackoverflow pages in the past and have not found a solution that works in any maintainable way.  I'll accept the answer here for you.

**Reply by Robert-Adelard** [reply]
@cuddihyge The title of the StackOverflow article is misleading - it's actually about how to force the browser to reload a page. Adding a version number to the file name seems like a reasonable approach.

**Reply by cuddihyge** [reply]
Yes, that particular solution will not work on our stack.  But I've found something very similar and will try it in an upcoming dev release.  Thx.

**Comment by cuddihyge** [comment]
I've found a more maintainable version of this solution that works with require.js on our tech stack.  It has been pushed to the SemTK 20210630 release.
